core/src/main/java/io/micronaut/core/io/service/MicronautMetaServiceLoaderUtils.java

```diff
@@ -0,0 +1,338 @@
+     * @param serviceClass The service class
+     * @param predicate    The predicate
+     * @param <S>          The service type
+     * @return the result
+     */
+    @NonNull
+    public static <S> List<S> findMetaMicronautServiceEntries(@NonNull Function<Integer, List<S>> constructor,
+                                                              @NonNull ClassLoader classLoader,
+                                                              @NonNull Class<S> serviceClass,
+                                                              @Nullable Predicate<S> predicate) {
+        SoftServiceLoader.StaticServiceLoader<S> staticServiceLoader = (SoftServiceLoader.StaticServiceLoader<S>) SoftServiceLoader.STATIC_SERVICES.get(serviceClass.getName());
+        if (staticServiceLoader != null) {
+            return staticServiceLoader.load(predicate);
+        }
+        return new MicronautServiceCollector<>(classLoader, serviceClass.getName(), predicate)
+            .collect(constructor, true);
+    }
+
+    /**
+     * Find Micronaut service entries.
+     *
+     * @param classLoader The classloader
+     * @param serviceName The service name
+     * @return The entries
+     * @throws IOException
+     */
+    @NonNull
+    public static Set<String> findMicronautMetaServiceEntries(@NonNull ClassLoader classLoader,
+                                                              @NonNull String serviceName) throws IOException {
+        if (cachedServices == null || cachedServicesClassloader != classLoader) {
+            cachedServices = findAllMicronautMetaServices(classLoader);
+            cachedServicesClassloader = classLoader;
+        }
+        return cachedServices.getOrDefault(serviceName, Set.of());
+    }
+
+    /**
+     * Find all Micronaut services.
+     *
+     * @param classLoader The classloader
+     * @return the all entries
+     * @throws IOException
+     */
+    @NonNull
+    public static Map<String, Set<String>> findAllMicronautMetaServices(@NonNull ClassLoader classLoader) throws IOException {
+        final ServiceScanner.StaticServiceDefinitions ssd = ServiceScanner.findStaticServiceDefinitions();
+        if (ssd != null) {
+            return ssd.serviceTypeMap();
+        }
+        List<URI> resourceDefs = IOUtils.getResources(classLoader, MICRONAUT_SERVICES_PATH);
+        if (resourceDefs.isEmpty()) {
+            return Map.of();
+        }
+
+        Map<String, Set<String>> services = new LinkedHashMap<>();
+
+        FileVisitor<Path> visitor = new FileVisitor<>() {
+
+            private Set<String> definitions;
+
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+                if (dir.endsWith(MICRONAUT_SERVICES_PATH)) {
+                    return FileVisitResult.CONTINUE;
+                }
+                String serviceName = dir.getFileName().toString();
+                definitions = services.get(serviceName);
```

Comment:
intellij tells me computeIfAbsent is possible here.

Reply:
I better to avoid lambda in the init phase